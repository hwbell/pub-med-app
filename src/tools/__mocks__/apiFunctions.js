const fetch = require('node-fetch')

// query and return based on date / citations
export async function getArticles(query, sortParam) {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  // let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;

  // let headers = {}
  // headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let list = [];
  for (let i = 0; i < 50; i++) {
    list.push({
      title: `Article ${i + 1}`,
      pmid: `PMID10000${i + 1}`
    })
  }

  let articles = await Promise.resolve({
    resultList: {
      result: list
    }
  })

  return articles;
}

// parses the result of a search query to a simple array of the article titles
export function parseSearchToTitlesArray(searchResults) {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })

  return titlesArray;
}