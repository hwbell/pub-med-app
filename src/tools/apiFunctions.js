const fetch = require('node-fetch')

// query and return based on date / citations
const getArticles = async (query, sortParam) => {

  // must have a query
  if (!query) return {};

  let allowedSorters = [
    '%20sort_cited:y',
    '%20sort_date:y',
  ];

  // check for bad strings and default to no sorter, which will sort by relevance
  let sorter;
  if (!sortParam) {
    sorter = '';
  } else {
    sorter = `%20sort_${sortParam}:y`;
    if (!allowedSorters.indexOf(sorter)) {
      sorter = '';
    }
  }

  let searchUrl = `https://www.ebi.ac.uk/europepmc/webservices/rest/searchPOST?query=${query}${sorter}&resultType=core&pageSize=10&format=json`;
  
  let headers = {}
  headers['Content-Type'] = 'application/x-www-form-urlencoded';

  let articles = await fetch(searchUrl, {
    method: 'POST',
    headers
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return articles;
}

// parses the result of a search query to a simple array of the article titles
const parseSearchToTitlesArray = (searchResults) => {
  let titlesArray = searchResults.resultList.result.map((result) => {
    return result.title;
  })
  return titlesArray;
}

// let results = getArticles('cancer').then((response) => {
//   console.log(response.resultList.result.slice(0,2))
// })

module.exports = {
  getArticles,
  parseSearchToTitlesArray,
}
