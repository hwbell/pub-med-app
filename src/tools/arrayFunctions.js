// this function will remove an article from a specific collection in the collections array
const removeArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles = collection.articles.filter(item => item.id !== article.id)
    }

  });

  return collections;
}

// this function will add an article to a specific collection
const addArticle = (collections, collectionName, article) => {

  // find the matching collection and remove the article
  collections.forEach((collection, i) => {

    if (collection.name === collectionName) {
      collection.articles.push(article);
    }

  });

  return collections;
}

module.exports = {
  removeArticle,
  addArticle
}