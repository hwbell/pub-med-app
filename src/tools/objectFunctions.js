
// this comibines two objects for the keys specified, giving priority to obj1
const combineObjects = (obj1, obj2, keys) => {

  if(!obj1 || !obj2) {
    return console.log('object argument missing')
  }
  if(!keys || !keys.length) {
    return console.log('keys argument missing')
  }

  let combined = {};

  keys.forEach((key) => {
    if (obj1[key]) {
      combined[key] = obj1[key];
    } else if (obj2[key]){
      combined[key] = obj2[key];
    } else {
      combined[key] = '';
    }
  })

  return combined;
}

module.exports = {
  combineObjects
}