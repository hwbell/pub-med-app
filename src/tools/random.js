let list = [];
for (let i = 0; i < 50; i++) {
  list.push({
    title: `Article ${i + 1}`,
    pmid: `PMID10000${i}`
  })
}

console.log(list)