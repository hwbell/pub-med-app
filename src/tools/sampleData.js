
// set some sample data for tests

export function getSampleCollections() {
  let collections = [];
  for (let i = 0; i < 10; i++) {
    collections.push({
      name: `collection ${i + 1}`,
      articles: [
        {
          name: `article ${i + 1}`,
          name: `article ${i + 2}`,
          name: `article ${i + 3}`,
        }
      ]
    })
  }
  return collections;
}