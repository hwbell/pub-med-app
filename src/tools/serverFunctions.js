const collectionServerUrl = process.env.REACT_APP_COLLECTION_SERVER_URL || 'http://localhost:3000/';
const fetch = require('node-fetch');

// this function will actually make a new user OR sign in an existing user, depending on the
// context
export async function signInUser(user, isNewUser) {
  if (!user) {
    return;
  }

  let headers = {}
  headers['content-type'] = 'application/application/json';

  let url = isNewUser ? `${collectionServerUrl}users` : `${collectionServerUrl}users/login`;
  let response = await fetch(url, {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
    },
  })
    .then(response => response.json())
    .then((response) => {
      // console.log(response)
      return response;
    })
    .catch(err => console.log(err))

  return response;
}

// patches the user's profile
export async function patchUser(patch, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}users/me`, {
    method: 'PATCH',
    body: JSON.stringify(patch),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// saves a collection to the server, or patches an exisiting collection on the server
export async function saveCollection(collection, headers, isExisting) {

  // console.log(collection)
  // assign the request params accoring to either PATCH for existing collections or 
  // POST for new collections

  // base url
  let url = `${collectionServerUrl}collections`;
  let method, body;

  if (isExisting) {
    // if its a patch, we only want to send allowed properties.
    let { name, articles } = collection;
    body = {
      name,
      articles
    }
    method = 'PATCH';
    url += `/${collection._id}`;
  } else {
    body = collection;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// deletes user's collection 
export async function deleteCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection._id}`, {
    method: 'DELETE',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// gets a user's profile / collections / threads
export async function getUserCollections(headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/me`, {
    method: 'GET',
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// patches an existing collection
export async function patchCollection(collection, headers) {
  let serverResponse = await fetch(`${collectionServerUrl}collections/${collection.id}`, {
    method: 'PATCH',
    body: JSON.stringify(collection),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      // console.log(response)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}

// get the threads on the server, 10 at a time, according to the sortBy param
export async function getPublicThreads(headers, sortBy, page) {

  if (!sortBy || !page) {
    return console.log('Missing sortBy or page parameter.');
  }
  // page represents results 1-10, 11-20, 21-30, etc. 
  let url = `${collectionServerUrl}threads/all/${sortBy}/${page}`;
  let method = 'GET';

  let serverResponse = await fetch(url, {
    method,
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}



// saves a collection to the server, or patches an exisiting collection on the server
export async function saveThread(thread, headers, isComment) {

  // assign the request params accoring to either PATCH for existing thread or 
  // POST for new thread

  // base url
  let url = `${collectionServerUrl}threads`;
  let method, body;

  if (isComment) {
    // if its a patch, we only want to send allowed properties.
    body = {
      user: thread.user,
      text: thread.text,
      add: thread.add
    };
    method = 'PATCH';
    url += `/comments/${thread._id}`;

  } else {
    body = thread;
    method = 'POST';
  }

  let serverResponse = await fetch(url, {
    method,
    body: JSON.stringify(body),
    headers
  })
    .then(response => response.json())
    .then((json) => {
      console.log(json)
      return json;
    })
    .catch(err => console.log(err))

  return serverResponse;
}








// let user = {
//   "name": makeRandomString(6),
//   "password": makeRandomString(8),
//   "email": `${makeRandomString(6)}@example.com`,
//   "age": 25
// }
// let collection = {
//   "name": "articles about science",
//   "articles": [
//     {
//       "url": "http://nature/bio/moggetarticle1"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle2"
//     },
//     {
//       "url": "http://nature/bio/moggetarticle3"
//     }
//   ]
// }

// let signInAttempt = signInUser(user).then((response) => {

//   console.log(localStorage.token)
// })

// let headers = {
//   Authorization: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI1ZDFhN2VlMTkzYmI0ZTQyZWM1Mjk5MmQiLCJpYXQiOjE1NjIwMTc1MDV9.AT5W-Vu8LssYs3fmH3IpnWSbQFxS0zk8S2WdW-2OAcs',
//   Accept: 'application/json',
//   'Content-Type': 'application/json',
// }
// let saveCollectionAttempt = saveCollection(collection, headers).then((response) => {
//   // 
//   console.log(response);
// })


// module.exports = {
//   signInUser,
//   saveCollection
// }