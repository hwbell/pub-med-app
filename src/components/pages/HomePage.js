import React from 'react';
import '../../App.scss';
import 'bootstrap/dist/css/bootstrap.css';

// components
import Header from '../Header';
import TextBlock from '../TextBlock';

// elements

// styling
// import ReactCSSTransitionGroup from 'react-addons-css-transition-group';

// tools
import { getArticles, parseSearchToTitlesArray } from '../../tools/apiFunctions.js';

// vars etc
const aboutIntro = {
  title: 'Welcome!',
  paragraph: [
    `PubMed Central (PMC) is a free digital repository that archives 
  publicly accessible full-text scholarly articles that have been published within the 
  biomedical and life sciences journal literature. This website is a portal to the resources
  provided to the public by PMC. Hopefully, this site will enable researchers to more efficiently
  organize, share and discuss scientific and medical literature. `,
  ],
  button: 'read more'
}

const resourcesIntro = {
  title: 'Resources',
  paragraph: [
    'abstracts > 35 million',
    'full text articles > 5 million',
    'patents > 4 million',
    'preprints > 80,000',
    'NIH clinical guidelines > 800'
  ],
  button: 'find resources!'
}

const journalsIntro = {
  title: 'Journals',
  paragraph: [
    'Nature',
    'Science',
    'Anatomy and Cell Biology',
    'PLoS Genetics',
    'Frontiers in Chemistry',
    'PNAS'
  ],
  button: 'search the literature!'
}

const toolsIntro = {
  title: 'Research',
  paragraph: [
    `Use this site to mine literature, organize resources, and connect with other reseachers! Using the tools
    found here, you can:`,
    `   - Quickly compile lists of important publications and research.`,
    `   - Connect with other researchers in a public forum to share thoughts and dscuss the latest scientific findings.`,
    `   - Hopefully enjoy the process of literature mining a little more than you used to!`
  ],
  button: 'start mining!'
}

class HomePage extends React.Component {
  constructor(props) {
    super(props);

    // this.fetchArticles = this.fetchArticles.bind(this);

    this.state = {
      newPublicationsInfo: {
        title: 'New Publications',
        paragraph: [],
        button: 'more articles'
      }
    };
  }

  render() {
    return (
      <div className="page">

        {/* the header */}
        <div className="glass">
        <Header
          title={"PubMed Central"}
          subtitle={"making research accessible"}
        />
        </div>

        <div className="row no-gutters row-eq-height">

          <div className="col col-12 col-sm-8">
            <TextBlock
              buttons={[{ link: '/about', text: 'read more' }]}
              text={aboutIntro}
            />
          </div>

          <div className="col col-12 col-sm-4">
            <TextBlock
              buttons={[{ link: '/search', text: 'find resources' }]}
              text={resourcesIntro}
            />
          </div>
        </div>

        <div className="row no-gutters row-eq-height">

          <div className="col col-12 col-sm-4">
            <TextBlock
              buttons={[{ link: '/about', text: 'search the literature' }]}
              text={journalsIntro}
            />
          </div>

          <div className="col col-12 col-sm-8">
            {this.state.newPublicationsInfo &&
              <TextBlock
                buttons={[{ link: '/search', text: 'start mining' }, { link: '/profile', text: 'register' }, { link: '/about', text: 'learn more' }]}
                text={toolsIntro}
              />}
          </div>
        </div>




      </div>
    );
  }
}

const styles = {
  main: {
    width: '100%'
  }
}

export default HomePage;
