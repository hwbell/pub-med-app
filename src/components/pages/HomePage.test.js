import React from 'react';
import ReactDOM from 'react-dom';
import HomePage from './HomePage';
import { shallow, mount } from 'enzyme';
import renderer from 'react-test-renderer';

// tools
import { getArticles, parseSearchToTitlesArray } from '../../tools/apiFunctions';

const DELAY_MS = 2000

const sleep = (ms) => {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// tests
it('renders without crashing', async () => {
  let wrapper = shallow(<HomePage />);
  // wrapper.update();

});

it('renders correctly', async () => {
  const tree = await renderer
    .create(<HomePage />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('contains the .page div', () => {
  let wrapper = shallow(<HomePage />);
  expect(wrapper.html()).toContain('div class=\"page\"')
});

it('contains Header elements', async () => {
  let wrapper = mount(<HomePage />);
  expect(wrapper.html()).toContain('h2 class=\"title\"');
  expect(wrapper.html()).toContain('h4 class=\"subtitle\"');
})

// this also tests that the TextBlock elements are getting rendered
it('contains .glass elements', async () => {
  let wrapper = mount(<HomePage />);
  expect(wrapper.find('.glass').length).toEqual(5);
})

it('contains search results after fetch', async () => {
  let wrapper = await shallow(<HomePage />)

  // should be empty to start
  expect(wrapper.state().newPublicationsInfo.paragraph.length).toBe(0)

  // then trigger the fetch
  await wrapper.instance().fetchArticles();
  
  // unfortunately it seems this is necessary? tried to find alternative, will
  // try again
  // await sleep(DELAY_MS)
  
  // should have results
  await wrapper.update();
  expect(wrapper.state().newPublicationsInfo.paragraph.length).not.toBe(0)
});

