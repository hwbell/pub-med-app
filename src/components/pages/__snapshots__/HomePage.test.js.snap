// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="page"
>
  <div
    className="glass"
  >
    <div
      className="header left-all-col"
    >
      <h2
        className="title"
      >
        PubMed Central
      </h2>
      <h4
        className="subtitle"
        style={
          Object {
            "alignSelf": "center",
          }
        }
      >
        making research accessible
      </h4>
    </div>
  </div>
  <div
    className="row no-gutters row-eq-height"
  >
    <div
      className="col col-12 col-sm-8"
    >
      <div
        className="glass"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
          }
        }
      >
        <p
          className="thread-title"
        >
          Welcome!
        </p>
        <div>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            PubMed Central (PMC) is a free digital repository that archives 
  publicly accessible full-text scholarly articles that have been published within the 
  biomedical and life sciences journal literature. This website is a portal to the resources
  provided to the public by PMC. Hopefully, this site will enable researchers to more efficiently
  organize, share and discuss scientific and medical literature. 
          </p>
        </div>
        <div
          style={
            Object {
              "margin": "0px 14px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/about"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              read more
            </a>
          </button>
        </div>
      </div>
    </div>
    <div
      className="col col-12 col-sm-4"
    >
      <div
        className="glass"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
          }
        }
      >
        <p
          className="thread-title"
        >
          Resources
        </p>
        <div>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            abstracts &gt; 35 million
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            full text articles &gt; 5 million
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            patents &gt; 4 million
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            preprints &gt; 80,000
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            NIH clinical guidelines &gt; 800
          </p>
        </div>
        <div
          style={
            Object {
              "margin": "0px 14px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/search"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              find resources
            </a>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row no-gutters row-eq-height"
  >
    <div
      className="col col-12 col-sm-4"
    >
      <div
        className="glass"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
          }
        }
      >
        <p
          className="thread-title"
        >
          Journals
        </p>
        <div>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            Nature
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            Science
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            Anatomy and Cell Biology
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            PLoS Genetics
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            Frontiers in Chemistry
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            PNAS
          </p>
        </div>
        <div
          style={
            Object {
              "margin": "0px 14px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/search"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              search the literature
            </a>
          </button>
        </div>
      </div>
    </div>
    <div
      className="col col-12 col-sm-8"
    >
      <div
        className="glass"
        style={
          Object {
            "alignItems": "flex-start",
            "display": "flex",
            "flexDirection": "column",
            "height": "100%",
            "justifyContent": "space-between",
          }
        }
      >
        <p
          className="thread-title"
        >
          Research
        </p>
        <div>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
            Use this site to mine literature, organize resources, and connect with other reseachers! Using the tools
    found here, you can:
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
               - Quickly compile lists of important publications and research.
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
               - Connect with other researchers in a public forum to share thoughts and dscuss the latest scientific findings.
          </p>
          <p
            className="thread-text"
            style={
              Object {
                "alignSelf": "flex-start",
              }
            }
          >
               - Hopefully enjoy the process of literature mining a little more than you used to!
          </p>
        </div>
        <div
          style={
            Object {
              "margin": "0px 14px",
            }
          }
        >
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/search"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              start mining
            </a>
          </button>
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/profile"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              register
            </a>
          </button>
          <button
            aria-label={null}
            className="article-button btn btn-primary btn-sm"
            onClick={[Function]}
          >
            <a
              href="/about"
              onClick={[Function]}
              style={
                Object {
                  "color": "white",
                }
              }
            >
              learn more
            </a>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
